LinearDecayCorrelationFunction.cpp

```diff
@@ -0,0 +1,189 @@
+//    parameters in a community sensor model (CSM).
+//
+//    LIMITATIONS:       None
+//
+//
+//    SOFTWARE HISTORY:
+//     Date          Author   Comment
+//     -----------   ------   -------
+//     01-Dec-2021   JPK      Adapted from LinearDecayCorrelationModel
+//
+//    NOTES:
+//     Refer to LinearDecayCorrelationFunction.h for more information.
+//#############################################################################
+#define CSM_LIBRARY
+#include "LinearDecayCorrelationFunction.h"
+#include "Error.h"
+
+#include <cmath>
+
+namespace csm {
+
+LinearDecayCorrelationFunction::LinearDecayCorrelationFunction()
+   :
+      SPDCorrelationFunction ("LinearDecayCorrelation"),
+      theCorrParams          ()
+{}
+
+LinearDecayCorrelationFunction::
+LinearDecayCorrelationFunction(const std::vector<double>& initialCorrsPerSegment,
+                               const std::vector<double>& timesPerSegment)
+   :
+      SPDCorrelationFunction ("LinearDecayCorrelation"),
+      theCorrParams          ()
+{
+   setCorrelationParameters(initialCorrsPerSegment,timesPerSegment);
+}
+
+LinearDecayCorrelationFunction::
+LinearDecayCorrelationFunction(const LinearDecayCorrelationModel::
+                                     Parameters& corrParams)
+   :
+      SPDCorrelationFunction ("LinearDecayCorrelation"),
+      theCorrParams          (corrParams)
+{
+   setCorrelationParameters(corrParams);
+}
+
+LinearDecayCorrelationFunction::~LinearDecayCorrelationFunction()
+{
+}
+
+void LinearDecayCorrelationFunction::setCorrelationParameters(
+   const std::vector<double>& initialCorrsPerSegment,
+   const std::vector<double>& timesPerSegment)
+{
+   setCorrelationParameters(LinearDecayCorrelationModel::
+                            Parameters(initialCorrsPerSegment,
+                                       timesPerSegment));
+}
+
+void
+LinearDecayCorrelationFunction::
+setCorrelationParameters(const LinearDecayCorrelationModel::Parameters& params)
+{
+   checkParameters(params);
+   
+   // store the correlation parameter values
+   theCorrParams = params;
+}
+
+double
+LinearDecayCorrelationFunction::getCorrelationCoefficient(double deltaTime) const
+{
+   return correlationCoefficientFor(theCorrParams,deltaTime);
+}
+
+double LinearDecayCorrelationFunction::
+       correlationCoefficientFor(const LinearDecayCorrelationModel::
+                                       Parameters& params,
+                                 double            deltaTime)
+{
+   const size_t size = params.theInitialCorrsPerSegment.size();
+   
+   if (size == 0) return 0.0;
+   
+   const double adt = fabs(deltaTime);
+   double prevCorr = params.theInitialCorrsPerSegment[0];
+   double prevTime = params.theTimesPerSegment[0];
+
+   double correlation = prevCorr;
+
+   for(size_t s = 1; s < size; ++s)
+   {
+      const double corr = params.theInitialCorrsPerSegment[s];
+      const double time = params.theTimesPerSegment[s];
```

Comment:
There is also a system call named `time()`, so naming this variable `time` could be confusing.

Reply:
Renamed  "corr" to "currCorr" and "time" to "currTime"